### PR TITLE
[Finder] added "use natural sort" option

### PR DESCRIPTION
**Pull request by vyshkant**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #26930
| License       | MIT
| Doc PR        | symfony/symfony-docs#9671

Added `$useNaturalSort` optional argument to `Finder::sortByName()` method. If it is specified and equals to `true`, ["natural sort order" algorithm](https://en.wikipedia.org/wiki/Natural_sort_order) will be applied, which means that `strnatcmp` function will be used instead of `strcmp` (see #26930 for details).

### CODE REVIEW

File: src/Symfony/Component/Finder/Finder.php [hunk 1]
```diff
@@ -397,13 +397,17 @@ public function sort(\Closure $closure)
...
      * This can be slow as all the matching files and directories must be retrieved for comparison.
      *
+     * @param bool $naturalOrdering Whether to use natural ordering or not
+     *
      * @return $this
      *
      * @see SortableIterator
      */
-    public function sortByName()
+    public function sortByName($naturalOrdering = false)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Finder.php]: for BC concerns, this should be handled using `func_num_args()/func_get_arg(0)` (adding a new argument breaks child classes). 
The signature should then be written as `sortByName(/* bool $useNaturalOrdering = false */)`

File: src/Symfony/Component/Finder/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * deprecated `Symfony\Component\Finder\Iterator\FilterIterator`
  * added Finder::hasResults() method to check if any results were found
+ * added $naturalOrdering option to Finder::sortByName() method
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/CHANGELOG.md]: `$useNaturalOrdering` looks more explicit to me

File: src/Symfony/Component/Finder/Finder.php [hunk 3]
```diff
@@ -397,13 +397,17 @@ public function sort(\Closure $closure)
      *
      * This can be slow as all the matching files and directories must be retrieved for comparison.
      *
+     * @param bool $naturalOrdering Whether to use natural ordering or not
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Finder.php]: `@param bool $useNaturalOrdering Whether to use natural ordering or not, disabled by default`

File: src/Symfony/Component/Finder/Finder.php [hunk 4]
```diff
@@ -397,13 +397,17 @@ public function sort(\Closure $closure)
...
+     *
      * @return $this
      *
      * @see SortableIterator
      */
-    public function sortByName()
+    public function sortByName($naturalOrdering = false)
     {
-        $this->sort = Iterator\SortableIterator::SORT_BY_NAME;
+        $this->sort = $naturalOrdering
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Finder.php]: our CS favors single lines, please adjust

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Almost good, thank you.

File: src/Symfony/Component/Finder/Finder.php (hunk 1 above)
Comment (javiereguiluz) [reply src/Symfony/Component/Finder/Finder.php]: Most of the times ["sort" and "order" are interchangeable](https://ell.stackexchange.com/questions/127765/are-sort-and-order-interchangeable). However, given that our methos is called `sortByName`, we use a `$this->sort` property and PHP calls this "natsort" for "natural sort" ... could we please rename `$useNaturalOrdering` to `$useNaturalSort` ?

**Comment by vyshkant** [comment]
> New features should target master (aka 4.2)!

@nicolas-grekas So I have to add a message to `CHANGELOG.md` under 4.2.0 section (not under 4.1.0)?

**Comment by nicolas-grekas** [comment]
> add a message to CHANGELOG.md under 4.2.0 section

yes please

File: src/Symfony/Component/Finder/Finder.php [hunk 5]
```diff
@@ -397,13 +397,17 @@ public function sort(\Closure $closure)
...
+     *
      * @return $this
      *
      * @see SortableIterator
      */
-    public function sortByName()
+    public function sortByName(/* bool $useNaturalSort = false */)
     {
-        $this->sort = Iterator\SortableIterator::SORT_BY_NAME;
+        $useNaturalSort = 1 === func_num_args() && true === func_get_arg(0);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Finder/Finder.php]: I'd be a more lax here: `0 < func_num_args() && func_get_arg(0)`

**Comment by vyshkant** [comment]
Does anybody have any idea why AppVeyor build has been failed?

**Review by nicolas-grekas — APPROVED** [review]
(failure unrelated)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @vyshkant.

**Pull request merged**